### PR TITLE
Moving async API import method to persistence layer

### PR DESCRIPTION
**Pull request by YasasRangika**
### Goals
To provide async API creation support for Choreo APIM core.[1]

### Objective
Move saveAsyncApiDefinition (registry) method to persistence layer

[1]: Issue: https://github.com/wso2-enterprise/choreo/issues/5567

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java [hunk 1]
```diff
@@ -1893,6 +1893,59 @@ public String getOASDefinition(Organization org, String apiId) throws OASPersist
         }
         return definition;
     }
+
+    @Override
+    public void saveAsyncDefinition(Organization org, String apiId, String apiDefinition)
+            throws AsyncSpecPersistenceException {
+        boolean isTenantFlowStarted = false;
+        try{
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java:1901]: Format

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java [hunk 2]
```diff
@@ -565,7 +565,7 @@ public PublisherAPI updateAPI(Organization org, PublisherAPI publisherAPI) throw
                     resource = registry.get(resourcePath);
                 }
                 resource.setContent(api.getSwaggerDefinition());
-                resource.setMediaType("application/json");
+                resource.setMediaType(APIConstants.API_ASYNCAPI_DEFINITION_RESOURCE_NAME);
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java:568]: Wrong constant

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java [hunk 3]
```diff
@@ -1833,7 +1833,7 @@ public void saveOASDefinition(Organization org, String apiId, String apiDefiniti
                 resource = registry.get(resourcePath);
             }
             resource.setContent(apiDefinition);
-            resource.setMediaType("application/json");
+            resource.setMediaType(APIConstants.API_ASYNCAPI_DEFINITION_RESOURCE_NAME);
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java:1836]: same here

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/ImportUtils.java [hunk 4]
```diff
@@ -237,7 +237,8 @@ public static API importApi(String extractedFolderPath, APIDTO importedApiDTO, B
                 }
                 targetApi.setOrganization(organization);
                 importedApi = PublisherCommonUtils
-                        .updateApi(targetApi, importedApiDTO, RestApiCommonUtil.getLoggedInUserProvider(), tokenScopes);
+                        .updateApi(targetApi, importedApiDTO, RestApiCommonUtil.getLoggedInUserProvider(), tokenScopes,
+                                organization);
             } else {
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/ImportUtils.java:242]: No need to pass organization . we could use targetApi.getOrganization()

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java [hunk 5]
```diff
@@ -114,7 +114,8 @@
      * @throws APIManagementException If an error occurs while updating the API
      * @throws FaultGatewaysException If an error occurs while updating manage of an existing API
      */
-    public static API updateApi(API originalAPI, APIDTO apiDtoToUpdate, APIProvider apiProvider, String[] tokenScopes)
+    public static API updateApi(API originalAPI, APIDTO apiDtoToUpdate, APIProvider apiProvider, String[] tokenScopes,
+            String organization)
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java:118]: We don't need separate organization param. We can use organization from originalAPI obj

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java [hunk 6]
```diff
@@ -785,7 +785,7 @@ public Response updateAPI(String apiId, APIDTO body, String ifMatch, MessageCont
             originalAPI.setOrganization(organization);
             //validate API update operation permitted based on the LC state
             validateAPIOperationsPerLC(originalAPI.getStatus());
-            API updatedApi = PublisherCommonUtils.updateApi(originalAPI, body, apiProvider, tokenScopes);
+            API updatedApi = PublisherCommonUtils.updateApi(originalAPI, body, apiProvider, tokenScopes, organization);
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java:788]: we don't need organization here as well

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java [hunk 7]
```diff
@@ -231,7 +231,7 @@ public PublisherAPI addAPI(Organization org, PublisherAPI publisherAPI) throws A
                     resource = registry.get(resourcePath);
                 }
                 resource.setContent(api.getSwaggerDefinition());
-                resource.setMediaType("application/json");
+                resource.setMediaType(APIConstants.API_ASYNCAPI_DEFINITION_RESOURCE_NAME);
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java:234]: wrong constant

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/APIPersistence.java [hunk 8]
```diff
@@ -255,6 +255,16 @@ DevPortalContentSearchResult searchContentForDevPortal(Organization org, String
     /* ==== Async API Definition ==========
     ============================================= */
 
+    /**
+     * Save OAS Schema definition
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/APIPersistence.java:259]: Change to 'Save Async API definition'

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/APIPersistence.java [hunk 9]
```diff
@@ -255,6 +255,16 @@ DevPortalContentSearchResult searchContentForDevPortal(Organization org, String
...
+     *
+     * @param org           Organization the Async API definition is owned by
+     * @param apiId         API ID
+     * @param apiDefinition API OAS definition
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/APIPersistence.java:263]: OAS --> Async API

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 10]
```diff
@@ -7831,6 +7832,27 @@ public void saveAsyncApiDefinition(API api, String jsonText) throws APIManagemen
         }
     }
 
+    public void saveAsyncApiDefinition(API api, String jsonText, String organization) throws APIManagementException {
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:7835]: Let's use the existing saveAsyncApiDefinition() method and pass organization instead

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java (hunk 1 above)
Comment (YasasRangika) [reply components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10621/commits/d5abe94f46e9e929b22ad4e684ae4d9122edc925

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java (hunk 2 above)
Comment (YasasRangika) [reply components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10621/commits/d5abe94f46e9e929b22ad4e684ae4d9122edc925

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java (hunk 3 above)
Comment (YasasRangika) [reply components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10621/commits/d5abe94f46e9e929b22ad4e684ae4d9122edc925

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java (hunk 7 above)
Comment (YasasRangika) [reply components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10621/commits/d5abe94f46e9e929b22ad4e684ae4d9122edc925

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/APIPersistence.java (hunk 8 above)
Comment (YasasRangika) [reply components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/APIPersistence.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10621/commits/d5abe94f46e9e929b22ad4e684ae4d9122edc925

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/APIPersistence.java (hunk 9 above)
Comment (YasasRangika) [reply components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/APIPersistence.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10621/commits/d5abe94f46e9e929b22ad4e684ae4d9122edc925

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 10 above)
Comment (YasasRangika) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10621/commits/d5abe94f46e9e929b22ad4e684ae4d9122edc925

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/ImportUtils.java (hunk 4 above)
Comment (YasasRangika) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/ImportUtils.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10621/commits/d711021b49bdfd26cece08490dee2dcd69b5d626

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java (hunk 6 above)
Comment (YasasRangika) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10621/commits/d711021b49bdfd26cece08490dee2dcd69b5d626

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java (hunk 5 above)
Comment (YasasRangika) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10621/commits/d711021b49bdfd26cece08490dee2dcd69b5d626

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java [hunk 11]
```diff
@@ -1893,6 +1893,62 @@ public String getOASDefinition(Organization org, String apiId) throws OASPersist
...
+            }
+
+            GenericArtifact apiArtifact = artifactManager.getGenericArtifact(apiId);
+            String apiProviderName = apiArtifact.getAttribute(APIConstants.API_OVERVIEW_PROVIDER);
+            String apiName = apiArtifact.getAttribute(APIConstants.API_OVERVIEW_NAME);
+            String apiVersion = apiArtifact.getAttribute(APIConstants.API_OVERVIEW_VERSION);
+            String visibility = apiArtifact.getAttribute(APIConstants.API_OVERVIEW_VISIBILITY);
+            String visibleRoles = apiArtifact.getAttribute(APIConstants.API_OVERVIEW_VISIBLE_ROLES);
+            String resourcePath = RegistryPersistenceUtil
+                    .getAsyncAPIDefinitionFilePath(apiName, apiVersion, apiProviderName);
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java:1921]: Let's use `String resourcePath = GovernanceUtils.getArtifactPath(registryType, apiId);` to get the artifact path.  see https://github.com/wso2/carbon-apimgt/blob/master/components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java#L1873

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/utils/RegistryPersistenceUtil.java [hunk 12]
```diff
@@ -1219,6 +1219,11 @@ public static String getOpenAPIDefinitionFilePath(String apiName, String apiVers
                 + RegistryConstants.PATH_SEPARATOR + apiName + RegistryConstants.PATH_SEPARATOR + apiVersion
                 + RegistryConstants.PATH_SEPARATOR;
     }
+
+    public static String getAsyncAPIDefinitionFilePath(String apiName, String apipVersion, String apiProvider) {
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/utils/RegistryPersistenceUtil.java:1223]: Let's not use this . use `String resourcePath = GovernanceUtils.getArtifactPath(registryType, apiId);` to get the path

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java [hunk 13]
```diff
@@ -759,14 +759,16 @@ public static API addAPIWithGeneratedSwaggerDefinition(APIDTO apiDto, String oas
...
         }
 
+        //adding the api
+        apiProvider.addAPI(apiToAdd);
+
+        apiToAdd.setOrganization(organization);
         if (isAsyncAPI) {
             AsyncApiParser asyncApiParser = new AsyncApiParser();
             String apiDefinition = asyncApiParser.generateAsyncAPIDefinition(apiToAdd);
             apiProvider.saveAsyncApiDefinition(apiToAdd, apiDefinition);
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java:769]: Let's remove this extra persistence layer call and save the definition from the addAPI() method. similar to what we do for openapi definition. See https://github.com/wso2/carbon-apimgt/blob/master/components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java#L223

Set the async definition to the API obj here and handle it from `apiProvider.addAPI(apiToAdd);`

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java [hunk 14]
```diff
@@ -759,14 +759,16 @@ public static API addAPIWithGeneratedSwaggerDefinition(APIDTO apiDto, String oas
...
         }
 
-        //adding the api
-        apiProvider.addAPI(apiToAdd);
         return apiToAdd;
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java:772]: Keep this

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java [hunk 15]
```diff
@@ -759,14 +759,16 @@ public static API addAPIWithGeneratedSwaggerDefinition(APIDTO apiDto, String oas
             apiToAdd.setAsyncApiDefinition(asyncApiDefinition);
         }
 
+        //adding the api
+        apiProvider.addAPI(apiToAdd);
+
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java:764]: Remove this. see the reason below

**Review by chamilaadhi — CHANGES_REQUESTED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java [hunk 16]
```diff
@@ -1761,6 +1761,7 @@ Mediation updateApiSpecificMediationPolicyContent(String apiId, Mediation mediat
      * @throws APIManagementException
      */
     void saveAsyncApiDefinition(API api, String jsonText) throws APIManagementException;
+
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java:1764]: Remove this empty line

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java (hunk 11 above)
Comment (YasasRangika) [reply components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java:1921]: Fixed in https://github.com/wso2/carbon-apimgt/pull/10626

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/utils/RegistryPersistenceUtil.java (hunk 12 above)
Comment (YasasRangika) [reply components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/utils/RegistryPersistenceUtil.java:1223]: Fixed in https://github.com/wso2/carbon-apimgt/pull/10626

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java (hunk 13 above)
Comment (YasasRangika) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java:769]: Fixed in https://github.com/wso2/carbon-apimgt/pull/10626

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java (hunk 15 above)
Comment (YasasRangika) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java:764]: Fixed in https://github.com/wso2/carbon-apimgt/pull/10626

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java (hunk 14 above)
Comment (YasasRangika) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java:772]: Fixed in https://github.com/wso2/carbon-apimgt/pull/10626

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java (hunk 16 above)
Comment (YasasRangika) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java:1764]: Fixed in https://github.com/wso2/carbon-apimgt/pull/10626

**Pull request closed without merging**